### PR TITLE
Encapsulation

### PR DESCRIPTION
**Pull request by dabrahams**

### CODE REVIEW

File: Sources/Compiler/Diagnostic.swift [hunk 1]
```diff
@@ -62,21 +62,27 @@ public struct Diagnostic: Hashable {
...
+  public static func error(
+    _ message: String, range: SourceRange? = nil,
+    children: [Diagnostic] = []
+  ) -> Diagnostic {
     Diagnostic(
       level: .error,
       message: message,
       location: range?.first(),
-      window: range.map({ r in Diagnostic.Window(range: r) }))
+      window: range.map({ r in Diagnostic.Window(range: r) }), children: children)
```
Comment (kyouko-taiga) [inline Sources/Compiler/Diagnostic.swift:73]:
```suggestion
      window: range.map({ r in Diagnostic.Window(range: r) }),
      children: children)
```

File: Sources/Compiler/Diagnostic.swift [hunk 2]
```diff
@@ -62,21 +62,27 @@ public struct Diagnostic: Hashable {
...
+  public static func warning(
+    _ message: String, range: SourceRange? = nil,
+    children: [Diagnostic] = []
+  ) -> Diagnostic {
     Diagnostic(
       level: .warning,
       message: message,
       location: range?.first(),
-      window: range.map({ r in Diagnostic.Window(range: r) }))
+      window: range.map({ r in Diagnostic.Window(range: r) }), children: children)
```
Comment (kyouko-taiga) [inline Sources/Compiler/Diagnostic.swift:85]:
```suggestion
      window: range.map({ r in Diagnostic.Window(range: r) }),
      children: children)
```

File: Sources/Compiler/Diagnostic.swift [hunk 3]
```diff
@@ -62,21 +62,27 @@ public struct Diagnostic: Hashable {
   }
 
   /// Creates an error diagnostic with `message` highlighting `range`.
-  public static func error(_ message: String, range: SourceRange? = nil) -> Diagnostic {
+  public static func error(
+    _ message: String, range: SourceRange? = nil,
+    children: [Diagnostic] = []
```
Comment (kyouko-taiga) [inline Sources/Compiler/Diagnostic.swift:67]:
```suggestion
    _ message: String,
    range: SourceRange? = nil,
    children: [Diagnostic] = []
```

File: Sources/Compiler/Diagnostic.swift [hunk 4]
```diff
@@ -62,21 +62,27 @@ public struct Diagnostic: Hashable {
...
       message: message,
       location: range?.first(),
-      window: range.map({ r in Diagnostic.Window(range: r) }))
+      window: range.map({ r in Diagnostic.Window(range: r) }), children: children)
   }
 
   /// Creates a warning diagnostic with `message` highlighting `range`.
-  public static func warning(_ message: String, range: SourceRange? = nil) -> Diagnostic {
+  public static func warning(
+    _ message: String, range: SourceRange? = nil,
```
Comment (kyouko-taiga) [inline Sources/Compiler/Diagnostic.swift:78]:
```suggestion
    _ message: String,
    range: SourceRange? = nil,
```

File: Sources/Compiler/TypeChecking/ConstraintGenerator.swift [hunk 5]
```diff
@@ -5,10 +5,16 @@ struct ConstraintGenerator: ExprVisitor {
 
   typealias Result = Void
 
-  /// Creates an instance for generating constraints in `scope`.
-  init(checker: TypeChecker, scope: AnyScopeID) {
+  /// Creates an instance, for use with `checker`, that generates constraints for `expr` in `scope`
+  /// with the given inferred and expected types.
+  init(
+    checker: TypeChecker, scope: AnyScopeID,
+    expr: AnyExprID, inferredType: Type?,  expectedType: Type?
```
Comment (kyouko-taiga) [inline Sources/Compiler/TypeChecking/ConstraintGenerator.swift:12]:
```suggestion
    checker: TypeChecker,
    scope: AnyScopeID,
    expr: AnyExprID,
    inferredType: Type?, 
    expectedType: Type?
```

File: Sources/Compiler/TypeChecking/TypeChecker.swift [hunk 6]
```diff
@@ -1396,17 +1400,19 @@ public struct TypeChecker {
 
     // Generate constraints.
     // Note: The constraint generator captures the ownership of `self`.
-    var generator = ConstraintGenerator(checker: self, scope: scope)
-    generator.inferredTypes[expr] = inferredType
-    generator.expectedTypes[expr] = expectedType
+    var generator = ConstraintGenerator(
+      checker: self, scope: scope, expr: expr,
+      inferredType: inferredType, expectedType: expectedType)
```
Comment (kyouko-taiga) [inline Sources/Compiler/TypeChecking/TypeChecker.swift:1405]:
```suggestion
      checker: self,
      scope: scope,
      expr: expr,
      inferredType: inferredType,
      expectedType: expectedType)
```

File: Sources/Compiler/TypeChecking/TypeChecker.swift [hunk 7]
```diff
@@ -1396,17 +1400,19 @@ public struct TypeChecker {
...
     constraints.append(contentsOf: generator.constraints)
 
     // Solve the constraints.
     var solver = ConstraintSolver(
-      checker: generator.checker.release(), scope: scope, fresh: constraints)
-    var solution = solver.solve()!
-    solution.diagnostics.append(contentsOf: generator.diagnostics)
+      checker: generator.checker.release(), scope: scope, fresh: constraints,
+      initialDiagnostics: generator.diagnostics
+    )
```
Comment (kyouko-taiga) [inline Sources/Compiler/TypeChecking/TypeChecker.swift:1414]:
```suggestion
      checker: generator.checker.release(),
      scope: scope,
      fresh: constraints,
      initialDiagnostics: generator.diagnostics)
```

**Review by kyouko-taiga — COMMENTED** [review]
Added suggestion to uphold my coding style.

File: Sources/Compiler/Diagnostic.swift (hunk 1 above)
Comment (dabrahams) [reply Sources/Compiler/Diagnostic.swift:73]: OK.  I do not believe in the “one-argument-per-line if the argument list is split” philosophy, but I can also see that without wrapping earlier lines it feels inconsistent.

BTW, [swift-format](https://github.com/apple/swift-format) implements my conventions, if you want to automate this stuff.

**Review by kyouko-taiga — APPROVED** [review]

**Pull request merged**